### PR TITLE
HNSW Refactor - benchmarks - [MOD-5371]

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

Update benchmarks to use new serialization files, made final decisions on the implementation of `processCandidate` and `processCandidate_Range`, uploaded a serialization script to the repo

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: tests/benchmark/data/hnsw_indices/hnsw_indices_all.txt [hunk 1]
```diff
@@ -1,14 +1,14 @@
-https://dev.cto.redis.s3.amazonaws.com/VectorSimilarity/DBpedia-n1M-cosine-d768-M64-EFC512.hnsw_v1
-https://dev.cto.redis.s3.amazonaws.com/VectorSimilarity/DBpedia-test_vectors-n10k.raw
+https://dev.cto.redis.s3.amazonaws.com/VectorSimilarity/dbpedia-cosine-dim768-M64-efc512.hnsw_v3
```
Comment (alonre24) [inline tests/benchmark/data/hnsw_indices/hnsw_indices_all.txt:1]: a special reason for omitting the dataset size?

File: tests/benchmark/data/serializer.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+        # Finished indexing, going back an indentation level so we close the input file
+
+        print('lastly, serializing the index')
+        output = os.path.join(location, serialized_file_name + '.hnsw_v3')
+        index.save_index(output)
+        print('saved to: ', output)
+
+        # Sanity check - attempt to load the index
+        index = HNSWIndex(output)
```
Comment (alonre24) [inline tests/benchmark/data/serializer.py:141]: Add a call for `checkIntegrity`, if we are already checking sanity

File: tests/benchmark/data/serializer.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
+import numpy as np
+from VecSim import *
+import h5py
+import os
+
+# location = os.path.join(os.path.abspath('.'), 'tests', 'benchmark', 'data', '')
+location = os.path.join(os.path.abspath('.'), '')
+print('at: ', location)
```
Comment (alonre24) [inline tests/benchmark/data/serializer.py:8]: Is this a leftover? 
Explain in a comment what should be the current directory when running this script

File: tests/benchmark/data/serializer.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        index.save_index(output)
+        print('saved to: ', output)
+
+        # Sanity check - attempt to load the index
+        index = HNSWIndex(output)
+
+        # Release memory before next iteration
+        del index
+
+if __name__ == '__main__':
```
Comment (alonre24) [inline tests/benchmark/data/serializer.py:146]: Great script
Consider adding some more explanations about its purpose and what it does at the top of this file and/or in the developer-benchmark docs...

File: tests/benchmark/data/hnsw_indices/ hnsw_indices_updated.txt [hunk 5]
```diff
@@ -1,2 +1,3 @@
-https://dev.cto.redis.s3.amazonaws.com/VectorSimilarity/DBpedia-n500K-cosine-d768-M65-EFC512.hnsw
-https://dev.cto.redis.s3.amazonaws.com/VectorSimilarity/DBpedia-n500K-cosine-d768-M65-EFC512-updated.hnsw
+https://dev.cto.redis.s3.amazonaws.com/VectorSimilarity/dbpedia-cosine-dim768-M65-efc512-n500k.hnsw_v3
+https://dev.cto.redis.s3.amazonaws.com/VectorSimilarity/dbpedia-cosine-dim768-M65-efc512-n500k-updated.hnsw_v3
+https://dev.cto.redis.s3.amazonaws.com/VectorSimilarity/dbpedia-cosine-dim768-test_vectors.raw
```
Comment (alonre24) [inline tests/benchmark/data/hnsw_indices/ hnsw_indices_updated.txt:3]: Is there a reason that we need the test vectors here now even though they weren't here before?

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 6]
```diff
@@ -580,11 +575,11 @@ void HNSWIndex<DataType, DistType>::emplaceToHeap(
...
 template <bool has_marked_deleted, typename Identifier>
 void HNSWIndex<DataType, DistType>::processCandidate(
-    idType curNodeId, ElementGraphData *cur_meta, const void *query_data, size_t layer, size_t ef,
-    tag_t *elements_tags, tag_t visited_tag,
-    vecsim_stl::abstract_priority_queue<DistType, Identifier> &top_candidates,
+    idType curNodeId, const void *query_data, size_t layer, size_t ef, tag_t *elements_tags,
+    tag_t visited_tag, vecsim_stl::abstract_priority_queue<DistType, Identifier> &top_candidates,
     candidatesMaxHeap<DistType> &candidate_set, DistType &lowerBound) const {
 
+    ElementGraphData *cur_meta = getGraphDataByInternalId(curNodeId);
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnsw.h:582]: meta? :)

File: tests/benchmark/data/hnsw_indices/hnsw_indices_all.txt (hunk 1 above)
Comment (GuyAv46) [reply tests/benchmark/data/hnsw_indices/hnsw_indices_all.txt:1]: No. I tried to make a standard naming template. We can add the data size to all the files, I tried to keep the most relevant parameters only (those you need to know in order to create the index - dim, type, metric, and M and efc)

File: tests/benchmark/data/serializer.py (hunk 2 above)
Comment (GuyAv46) [reply tests/benchmark/data/serializer.py:141]: Good call. Didn't know we have it

File: tests/benchmark/data/hnsw_indices/ hnsw_indices_updated.txt (hunk 5 above)
Comment (GuyAv46) [reply tests/benchmark/data/hnsw_indices/ hnsw_indices_updated.txt:3]: It makes more sense to have it here if one wishes to download these datasets only. On the `*_all.txt` it appears only once.

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 6 above)
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw.h:582]: :)
This PR is prior to the renaming one

File: tests/benchmark/data/serializer.py [hunk 7]
```diff
@@ -133,15 +162,22 @@ def serialize(files=DEFAULT_FILES):
...
         # Sanity check - attempt to load the index
         index = HNSWIndex(output)
+        index.check_integrity()
 
         # Release memory before next iteration
         del index
 
 if __name__ == '__main__':
-    serialize()
+    serialize([{
```
Comment (alonre24) [inline tests/benchmark/data/serializer.py:177]: Did you leave it like this (glove-25 only) on purpose?

**Review by alonre24 — APPROVED** [review]
Good comments!

File: tests/benchmark/data/serializer.py (hunk 7 above)
Comment (GuyAv46) [reply tests/benchmark/data/serializer.py:177]: nope. removing

**Pull request merged**